File: packages/strategies/src/rsi.ts
import { OrderSide, OrderType, ResearchMode } from '@quant/common';
import type { Bar } from '@quant/common';
import type { Strategy, StrategyContext, StrategyMeta, StrategyResult } from '@quant/strategy-runtime';
import { StrategyState } from '@quant/strategy-runtime';

export interface RSIParams {
  period: number;
  oversold: number;
  overbought: number;
}

/** RSI 策略 — RSI 低于超卖线买入，高于超买线卖出 */
export class RSIStrategy implements Strategy {
  readonly meta: StrategyMeta;
  private _state: StrategyState = StrategyState.Idle;
  private ctx: StrategyContext | null = null;
  private readonly closes: number[] = [];
  private readonly period: number;
  private readonly oversold: number;
  private readonly overbought: number;

  constructor(params: RSIParams = { period: 14, oversold: 30, overbought: 70 }) {
    this.period = params.period;
    this.oversold = params.oversold;
    this.overbought = params.overbought;
    this.meta = {
      name: 'rsi',
      description: 'RSI 策略：RSI 低于超卖线买入，高于超买线卖出',
      modes: [ResearchMode.Traditional],
      params: [
        { key: 'period', label: 'RSI 周期', type: 'number' as any, default: this.period, min: 2, max: 100 },
        { key: 'oversold', label: '超卖线', type: 'number' as any, default: this.oversold, min: 0, max: 50 },
        { key: 'overbought', label: '超买线', type: 'number' as any, default: this.overbought, min: 50, max: 100 },
      ],
      version: '1.0.0',
    };
  }

  get state() { return this._state; }

  init(context: StrategyContext): void {
    this.ctx = context;
    this._state = StrategyState.Running;
  }

  onBar(bar: Bar, context: StrategyContext): void {
    this.closes.push(bar.close);
    if (!this.ctx) return;

    const rsi = this.calculateRSI();
    if (rsi === null) return;

    const pos = context.getPosition(bar.symbol);

    // RSI 低于超卖线且无持仓 → 买入
    if (rsi < this.oversold && !pos) {
      context.submitOrder({
        symbol: bar.symbol,
        side: OrderSide.Buy,
        type: OrderType.Market,
        quantity: 100,
      });
    }

    // RSI 高于超买线且有持仓 → 卖出
    if (rsi > this.overbought && pos) {
      context.submitOrder({
        symbol: bar.symbol,
        side: OrderSide.Sell,
        type: OrderType.Market,
        quantity: pos.quantity,
      });
    }
  }

  finish(): StrategyResult {
    this._state = StrategyState.Stopped;
    return { meta: this.meta, orders: [], trades: [] };
  }

  /** 计算 RSI */
  private calculateRSI(): number | null {
    const changes: number[] = [];
    for (let i = 1; i < this.closes.length; i++) {
      changes.push(this.closes[i] - this.closes[i - 1]);
    }
    if (changes.length < this.period) return null;

    const recentChanges = changes.slice(-this.period);
    const gains = recentChanges.filter((c) => c > 0);
    const losses = recentChanges.filter((c) => c < 0).map((c) => Math.abs(c));

    const avgGain = gains.length > 0 ? gains.reduce((s, g) => s + g, 0) / this.period : 0;
    const avgLoss = losses.length > 0 ? losses.reduce((s, l) => s + l, 0) / this.period : 0;

    if (avgLoss === 0) return 100;
    const rs = avgGain / avgLoss;
    return 100 - 100 / (1 + rs);
  }
}

File: packages/strategies/quantforge_strategies/combined/rsi.py
"""RSI 策略"""

from __future__ import annotations

from collections import deque

from quantforge_strategy import (
    Strategy, StrategyMeta, StrategyResult, StrategyState,
    Bar, OrderSide, OrderType, OrderRequest, ParamType, ResearchMode,
    StrategyKind, StrategyParamDef, StrategyCategory, StrategySubcategory,
)
from ..indicators import rsi, last_valid


class RSIStrategy(Strategy):
    """RSI 策略：RSI < 30 买入，RSI > 70 卖出"""

    def __init__(self, period: int = 14, oversold: float = 30.0, overbought: float = 70.0) -> None:
        self._period = period
        self._oversold = oversold
        self._overbought = overbought
        self._prices: deque[float] = deque(maxlen=period + 2)
        self._bought = False

    @property
    def meta(self) -> StrategyMeta:
        return StrategyMeta(
            name="rsi",
            description="RSI 超买超卖策略",
            modes=[ResearchMode.Traditional],
            params=[
                StrategyParamDef(key="period", label="RSI 周期",
                                 type=ParamType.Number, default=self._period,
                                 min=2, max=50),
                StrategyParamDef(key="oversold", label="超卖阈值",
                                 type=ParamType.Number, default=self._oversold,
                                 min=10, max=50),
                StrategyParamDef(key="overbought", label="超买阈值",
                                 type=ParamType.Number, default=self._overbought,
                                 min=50, max=90),
            ],
            version="0.1.0",
            kind=StrategyKind.Timing,
            category=StrategyCategory.NON_FACTOR,
            subcategory=StrategySubcategory.TREND_CTA,
        )

    @property
    def state(self) -> StrategyState:
        return StrategyState.Idle

    def init(self, context) -> None:
        self._prices.clear()
        self._bought = False

    def on_bar(self, bar: Bar, context) -> None:
        self._prices.append(bar.close)

        if len(self._prices) < self._period + 1:
            return

        rsi_series = rsi(list(self._prices), self._period)
        rsi_value = last_valid(rsi_series)
        if rsi_value is None:
            return

        if rsi_value < self._oversold and not self._bought:
            account = context.get_account()
            qty = int(account.cash / bar.close)
            if qty > 0:
                context.submit_order(OrderRequest(
                    symbol=bar.symbol, side=OrderSide.Buy,
                    type=OrderType.Market, quantity=qty,
                    reason=f"RSI超卖：RSI={rsi_value:.2f}<{self._oversold}",
                ))
                self._bought = True
        elif rsi_value > self._overbought and self._bought:
            pos = context.get_position(bar.symbol)
            if pos and pos.quantity > 0:
                context.submit_order(OrderRequest(
                    symbol=bar.symbol, side=OrderSide.Sell,
                    type=OrderType.Market, quantity=int(pos.quantity),
                    reason=f"RSI超买：RSI={rsi_value:.2f}>{self._overbought}",
                ))
                self._bought = False

    def finish(self) -> StrategyResult:
        return StrategyResult(meta=self.meta)
